velox/row/CompactRow.cpp

```diff
@@ -0,0 +1,920 @@
+
+  //  <keys array> | <values array>
+
+  auto mapBase = decoded_.base()->asUnchecked<MapVector>();
+  auto offset = mapBase->offsetAt(baseIndex);
+  auto size = mapBase->sizeAt(baseIndex);
+
+  auto keysSerializedBytes = serializeAsArray(
+      children_[0], offset, size, childIsFixedWidth_[0], buffer);
+
+  auto valuesSerializedBytes = serializeAsArray(
+      children_[1],
+      offset,
+      size,
+      childIsFixedWidth_[1],
+      buffer + keysSerializedBytes);
+
+  return keysSerializedBytes + valuesSerializedBytes;
+}
+
+int32_t CompactRow::serialize(vector_size_t index, char* buffer) {
+  return serializeRow(index, buffer);
+}
+
+void CompactRow::serializeFixedWidth(vector_size_t index, char* buffer) {
+  VELOX_DCHECK(fixedWidthTypeKind_);
+  switch (typeKind_) {
+    case TypeKind::BOOLEAN:
+      *reinterpret_cast<bool*>(buffer) = decoded_.valueAt<bool>(index);
+      break;
+    case TypeKind::TIMESTAMP:
+      *reinterpret_cast<int64_t*>(buffer) =
+          decoded_.valueAt<Timestamp>(index).toMicros();
+      break;
+    default:
+      memcpy(
+          buffer,
+          decoded_.data<char>() + decoded_.index(index) * valueBytes_,
+          valueBytes_);
+  }
+}
+
+void CompactRow::serializeFixedWidth(
+    vector_size_t offset,
+    vector_size_t size,
+    char* buffer) {
+  VELOX_DCHECK(supportsBulkCopy_);
+  // decoded_.data<char>() can be null if all values are null.
+  if (decoded_.data<char>()) {
+    memcpy(
+        buffer,
+        decoded_.data<char>() + decoded_.index(offset) * valueBytes_,
+        valueBytes_ * size);
+  }
+}
+
+int32_t CompactRow::serializeVariableWidth(vector_size_t index, char* buffer) {
+  switch (typeKind_) {
+    case TypeKind::VARCHAR:
+      FOLLY_FALLTHROUGH;
+    case TypeKind::VARBINARY: {
+      auto value = decoded_.valueAt<StringView>(index);
+      writeInt32(buffer, value.size());
+      if (!value.empty()) {
+        memcpy(buffer + kSizeBytes, value.data(), value.size());
+      }
+      return kSizeBytes + value.size();
+    }
+    case TypeKind::ARRAY:
+      return serializeArray(index, buffer);
+    case TypeKind::MAP:
+      return serializeMap(index, buffer);
+    case TypeKind::ROW:
+      return serializeRow(index, buffer);
+    default:
+      VELOX_UNREACHABLE(
+          "Unexpected type kind: {}", mapTypeKindToName(typeKind_));
+  };
+}
+
+namespace {
+
+// Reads single fixed-width value from buffer and returns number of bytes read.
+// Stores the value into flatVector[index].
+template <typename T>
+size_t readFixedWidthValue(
+    bool isNull,
+    const char* buffer,
+    FlatVector<T>* flatVector,
+    vector_size_t index) {
+  if (isNull) {
+    flatVector->setNull(index, true);
+  } else if constexpr (std::is_same_v<T, Timestamp>) {
+    int64_t micros = *reinterpret_cast<const int64_t*>(buffer);
+    flatVector->set(index, Timestamp::fromMicros(micros));
+  } else {
+    T value = *reinterpret_cast<const T*>(buffer);
```

Comment:
`memory(&value, buffer, sizeof(T))` to be safe for unaligned memory

Reply:
https://blog.quarkslab.com/unaligned-accesses-in-cc-what-why-and-solutions-to-do-it-properly.html